File: 03_merge_two_sorted_lists.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        merged_list_head = None
        merged_list_curr = None
        while list1 or list2:
            node_to_append = None
            if list1 and list2:
                if list1.val < list2.val:
                    node_to_append = list1
                    list1 = list1.next
                else:
                    node_to_append = list2
                    list2 = list2.next
            elif list1:
                node_to_append = list1
                list1 = list1.next
            elif list2:
                node_to_append = list2
                list2 = list2.next

            if not merged_list_head:
                merged_list_head = node_to_append
                merged_list_curr = merged_list_head
            else:
                merged_list_curr.next = node_to_append
                merged_list_curr = merged_list_curr.next

        return merged_list_head

File: Java/03_merge_two_sorted_lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode merged_head = null;
        ListNode merged_curr = null;

        while (list1 != null || list2 != null) {
            ListNode appendNode = null;
            if (list1 != null && list2 != null) {
                if (list1.val < list2.val) {
                    appendNode = list1;
                    list1 = list1.next;
                } else {
                    appendNode = list2;
                    list2 = list2.next;
                }
            } else if (list1 != null) {
                appendNode = list1;
                list1 = list1.next;
            } else if (list2 != null) {
                appendNode = list2;
                list2 = list2.next;
            }
            if (merged_head == null) {
                merged_head = appendNode;
                merged_curr = merged_head;
            } else {
                merged_curr.next = appendNode;
                merged_curr = merged_curr.next;
            }
        }
        return merged_head;
    }
}